nixos/tpm2: init

lschuermann opened the pull request:
###### Motivation for this change
The TPM2 device files should be owned by a specific user. In addition to that, the TPM2 kernel resource manager (`/dev/tpmrm*`) should be owned by a specific group, such that an unprivileged user is able to access this device.

Furthermore, this commit adds a systemd service for the `tpm2-abrmd` userspace resource manager (merged in #72018), on which many applications still rely.


###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

@lassulus

lschuermann committed:
nixos/tpm2: init

This commit adds udev rules, the userspace resource manager and
PKCS#11 module support.

Infinisil commented on nixos/modules/security/tpm2.nix:
```diff
@@ -0,0 +1,120 @@
+  # the tssGroup is only allowed to access the kernel resource manager
+  # Therefore, if either of the two are null, the respective part isn't generated
+  udevRules = tssUser: tssGroup: ''
+    ${lib.optionalString (tssUser != null) ''KERNEL=="tpm[0-9]*", MODE="0660", OWNER="${tssUser}"''}
```
Since this function is only used once, I'd remove the arguments and use `cfg.tss{User,Group}` directly in it

lschuermann replied:
I think this function is complicated enough (or at least not very easy to read). Having user and group configurable, I'd _really_ like to insist on having this separate, even if it's just for reducing the visual complexity of the config section.

Infinisil commented on nixos/modules/security/tpm2.nix:
```diff
@@ -0,0 +1,120 @@
+          package = lib.mkOption {
+            description = "tpm2-abrmd package to use";
+            type = lib.types.package;
+            default = pkgs.tpm2-abrmd;
```
This needs `defaultText = "pkgs.tmp2-abrmd"` as well to render correctly in the manual

lschuermann replied:
Will do.

Infinisil commented on nixos/modules/security/tpm2.nix:
```diff
@@ -0,0 +1,120 @@
+              Set common environment variables to use the tpm2-abrmd resource manager
+              The variables are
+                - TPM2TOOLS_TCTI
+                - TPM2_PKCS11_TCTI
```
Needs xml madness for lists, see https://docbook.rocks/#lists

lschuermann replied:
Thanks for the hint.

Infinisil commented on nixos/modules/security/tpm2.nix:
```diff
@@ -0,0 +1,120 @@
+      services.udev.extraRules = lib.optionalString cfg.applyUdevRules (udevRules cfg.tssUser cfg.tssGroup);
+
+      users.users = lib.mkIf (cfg.tssUser != null) {
+        "${cfg.tssUser}" = {
```
You could do this as well, same for `users.groups`
```nix
users.users.${cfg.tssUser} = lib.mkIf (cfg.tssUser != null) {
  isSystemUser = true;
}
```

However usually such `user`/`group` options are of type `types.str` without allowing `null`, and the check is done as `mkIf (cfg.tssUser == "tss")` (meaning, only create this user in the module when the default value is used). I'd recommend doing that here

Infinisil commented on nixos/modules/security/tpm2.nix:
```diff
@@ -0,0 +1,120 @@
+    }
+    (lib.mkIf cfg.abrmd.enable {
+      systemd.services."tpm2-abrmd" = {
+        enable = true;
```
Can be removed as it's the default

infinisil reviewed (changes requested):
I don't know much about TPM, but I'm wondering whether it's worth having some of these options. In particular I can't see a use case for changing `tssUser` and `tssGroup`, a fixed value will probably work fine, same for `abrmd.user`. `applyUdevRules` seems also like it could be an unconfigurable `true`, would anybody want to turn that off? Same for `setEnvironmentVariables`. What do you think?

lschuermann commented:
@Infinisil Regarding your comments on the different (configurable) options:

- `tssUser` / `tssGroup`: I'd almost say that we could hardcode `tss` in here, but AFAIK TPM software is pretty fragile, and while I haven't seen an example where `tss` doesn't work yet, I'd rather be prepared. Configurable users / groups are the norm across NixOS modules. Also, AFAIK `tss` is more of a convention than a standard here, so I'd rather leave it that way.
- `applyUdevRules`: The `udev` rules are sensible, but quite opinionated (only the `tss` user can access the device, only the group the resource manager, permissions, etc.). I could imagine a few very valid scenarios where a user would opt out of those, but instead configure their own.
- `setEnvironmentVariables`: Just enabling the userspace resource manager should never influence the behaviour of TPM applications implicitly. Setting the environment variables doesn't work for all TPM tools (just tpm2-tools and tpm2-pkcs11 for now), and changes the default behaviour to always use the userspace resource manager. That is why it's in default - despite being quite useful - disabled. There are many other ways a user might want to interface their TPM(s), and with multiple they could even opt in to use different resource managers. Unsetting some environment variables is not a solution here.

I thereby conclude, that unfortunately the complexity of this module cannot be reduced significantly without trade offs.

Infinisil commented:
- `tss{User,Group}`: I think configurable user/group makes sense where there's a reason to change it, such as when the service creates files which need to be owned by the main system user. In this case however I can't see this being necessary. If at some later point we discover that this indeed needs to be configurable I'll be happy to merge additional options for it.
- `applyUdevRules`: Can you explain how you arrived at these udev rules? I don't quite get what they do, and also what `tmp` and `tmprm` are used for and when they're needed. I think usually permissions to devices are handled with a group users can add themselves to to allow them to use it, why is this not the case for `tmp` but for `tmprm`?
- `setEnvironmentVariables`: Alright that sounds reasonable. Though why should the rm never influence TPM applications implicitly?

lschuermann commented:
Extending my points above:

- `tss{User,Group}`: Actually, considering the udev rules, the files `/dev/tpm{[0-9],rm[0-9]}` will be owned by the user / group respectively. I have an application which would like exclusive ownership of the `/dev/tpm[0-9]` device, and as such I'd change the `tssUser` option to the application's user. I could hack my way around this, or even run the application as `tss` (not knowing what other implications that'll have), but changing device ownership appears to be the cleanest solution. Also, this way, other services could rely on the configured user/group. If you are still not convinced, I will remove the options. This implies, that with a hard-coded user and group, there must be a way to disable setting the udev rules.
- `applyUdevRules`: As clearly stated above the `udevRules` function, these are taken from the `tpm2-tss` package. Without going too much into detail, a TPM has limited resources, and the `/dev/tpm[0-9]` provides a vastly unmanaged interface to the TPM. As an exclusive owner of this file, you'd have full control and responsibility over resource management of the TPM. It therefore doesn't make sense to use groups here, so the group will be `root`. The `/dev/tpmrm[0-9]` is a resource manager implemented in the kernel. While not having full control, multiple application can simultaneously use the device without blocking each other. As this new (AFAIK kernel 4.x) device is not always supported, the userspace resource manager (using DBus as transport) should still be provided.
- `setEnvironmentVariables`: Each application may implement this differently, but most try to automatically detect the device to use. On my system, most try to use the `/dev/tpm0` device, although they don't have the required permissions. When using the userspace resource manager, and having only one TPM, this option provides an override for some TPM applications. It helps me, likely helps others, but should be opt-in as it is opinionated.

Infinisil commented:
Alright I see, you convinced me :)

Looks ready to merge then after the adjustments above

lschuermann commented:
Okay, so this has been a long time coming.

I've finally addressed your concerns @Infinisil and used this module in a working production setup, polishing one or the other thing. The behaviour now comes as close to "working out of the box" and "confirming to the TPM2 TSS desired setup" as I can think of.

The default configuration now leads to the following device ownership:
```

However, this module is now dependent on #72374 which should be ready too.
Let me know what you think.

lschuermann commented on nixos/modules/security/tpm2.nix:
```diff
@@ -0,0 +1,172 @@
+          enable = lib.mkEnableOption "TPM2 PKCS#11 tool in system path and shared library in /run/libtpm2_pkcs11.so";
+
+          librarySymlink = lib.mkOption {
+            description = "symlink the tpm2-pkcs11 shared library to /run/libtpm2_pkcs11.so";
```
I'm concerned about this:
To use it in application's configuration the PKCS#11 shared library should *really* be symlinked somewhere out of the Nix store. Does it make sense to do that as part of the activation script, and is there a better location?
Of course, you can turn the symlinking off.

Infinisil replied:
Why is this needed?

lschuermann replied:
For instance, when you want to use the PKCS#11 module for SSH, you'd put this in your config:
```
Host *
        PKCS11Provider /run/libtpm2_pkcs11.so
```
It would be very annoying to put an everchanging path from the nix store in there. I'd much rather use this module, which conveniently symlinks the current nixpkgs' version somewhere. Do you know of a better way to do this?

Infinisil replied:
How about using `programs.ssh.extraConfig` directly?

lschuermann replied:
I haven't tested this, but it may be an option. Regardless, the entire point of a PKCS#11 shared library is to be found and used by other applications - dynamically, not during build time. Often, configuration files require a path to the library. I should be exposed somewhere, preferably with a constant path pointing to the latest version. Is that something we support in NixOS, and how should I do this.

Infinisil replied:
Hmm I see. Okay then I think we can put this library into `environment.systemPackages` such that it's exposed at `/run/current-system/sw/lib/libtmp2_pkcs11.so`, and then drop the `librarySymlink` option because there shouldn't be a problem with always having it.

Infinisil replied:
Make sure to use `lib.getLib pkgs.tpm2-pkcs11` (see https://github.com/NixOS/nixpkgs/pull/72374#discussion_r372945734)

lschuermann replied:
That works flawlessly, thanks for the awesome hint!

lschuermann commented:
Now that #72374 is done, this one should be ready again. Could you remove the "blocked" label, @Infinisil?

infinisil reviewed (changes requested):
Still some things to improve, I hope you don't mind!

lschuermann commented:
There's always room for improvement - I'm willing to invest more time. Unfortunately, I got sick, hence the delay.

lschuermann commented on nixos/modules/security/tpm2.nix:
```diff
@@ -0,0 +1,162 @@
+      '';
+      type = lib.types.submodule ({
+        options = {
+          enable = lib.mkEnableOption "common TPM2 TCTI environment variables";
```
Can do.

Infinisil commented:
I saw some interest for this PR on IRC by @danderson, so here I am with a ping :)

lschuermann commented on nixos/modules/security/tpm2.nix:
```diff
@@ -0,0 +1,162 @@
+
+          name = lib.mkOption {
+            description = "The name of the TCTI, e.g. tabrmd or device";
+            type = with lib.types; nullOr (enum [ "tabrmd" "device" ] );
```
Yes, `null` is not a valid option here. The option tells the application what kind of interface to use when talking to the TPM, like using a D-Bus connection to some daemon or opening a device file. AFAIK this indicates a part of the name of a shared library which is then `dlopen()`ed. I tried improving the documentation.

lschuermann replied:
> Like do you need to set `abrmd.enable` if this is set to `tabrmd`?

Ultimately that would make sense in most scenarios, but I'd deliberately not couple the environment variables to the running daemon. Setting the environment variables is a method of overriding the automatic TCTI detection of many programs and I'm actively running the userspace resource manager for development while using the device (kernel resource manager) TCTI for everything else.

lschuermann commented on nixos/modules/security/tpm2.nix:
```diff
@@ -0,0 +1,162 @@
+          };
+
+          deviceConf = lib.mkOption {
+            description = "Configuration part of the device TCTI, e.g. the path to the TPM device";
```
I've inserted a link to the documentation for the format.

lschuermann commented on nixos/modules/security/tpm2.nix:
```diff
@@ -0,0 +1,162 @@
+      description = "TPM2 PKCS#11 module support";
+      type = lib.types.submodule ({
+        options = {
+          enable = lib.mkEnableOption "TPM2 PKCS#11 tool and shared library in system path";
```
Done.

lschuermann commented on nixos/modules/security/tpm2.nix:
```diff
@@ -0,0 +1,162 @@
+      # PKCS11 tools and library
+      environment.systemPackages = lib.optionals cfg.pkcs11.enable [ (lib.getBin cfg.pkcs11.package) (lib.getLib cfg.pkcs11.package) ];
+
+      services.udev.extraRules = lib.optionalString cfg.applyUdevRules (udevRules cfg.tssUser cfg.tssGroup);
```
Done.

lschuermann commented on nixos/modules/security/tpm2.nix:
```diff
@@ -0,0 +1,162 @@
+  config = lib.mkIf cfg.enable (lib.mkMerge [
+    {
+      # PKCS11 tools and library
+      environment.systemPackages = lib.optionals cfg.pkcs11.enable [ (lib.getBin cfg.pkcs11.package) (lib.getLib cfg.pkcs11.package) ];
```
Done.

lschuermann commented:
> I saw some interest for this PR on IRC by @danderson, so here I am with a ping :)

Glad to know there's interest. It'll be a big step if NixOS will support TPM2 modules without much hassle and manual setup. There's great potential, as long as the TPM is not an [Intel fTPM](https://blog.ptsecurity.com/2020/03/intelx86-root-of-trust-loss-of-trust.html).

I've tried to address all your change requests and they did improve the module quite a lot. Maybe this will be the last iteration? :) Feel free to "resolve" the conversations if you feel I've sufficiently addressed the underlying issues.

Infinisil commented on nixos/modules/security/tpm2.nix:
```diff
@@ -0,0 +1,175 @@
+
+    tssUser = lib.mkOption {
+      description = "Name of the tpm device-owner and service user, set if " +
+                    "applyUdevRules is set";
```
This formatting is very unconventional, you can just use
```nix
{
  description = ''
    Name of the ...
    applyUdevRules
  '';
}
```

Same for all the other places

Infinisil commented on nixos/modules/security/tpm2.nix:
```diff
@@ -0,0 +1,175 @@
+        default = false;
+      };
+
+      name = lib.mkOption {
```
A better name for this option would be "transmissionInterface" or just "interface"

Infinisil commented on nixos/modules/security/tpm2.nix:
```diff
@@ -0,0 +1,175 @@
+        (udevRules cfg.tssUser cfg.tssGroup);
+
+      # Create the tss user and group only if the default value is used
+      users.users.${cfg.tssUser} = lib.mkIf (cfg.tssUser == "tss") {};
```
Need to also set `isSystemUser = true;` here, and probably `group = cfg.tssGroup` (otherwise I don't think the systemd services runs under that group).

lschuermann replied:
Technically the systemd service doesn't need to run under any specific group, so I guess specifying `nogroup` is a good idea here.

Infinisil replied:
(The `isSystemUser` is still missing)

lschuermann replied:
Yes, totally forgot that, sorry.

lschuermann commented:
Ready for another review round I guess.

infinisil reviewed (approved):
Looking good!
